ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java

```diff
@@ -575,6 +579,113 @@ private JoinNode buildJoin(final Join root, final String prefix, final boolean i
+          throw new KsqlException("Invalid join type:"
+              + " full-outer join not supported for stream-table joins.");
+        }
+
+        if (joinOnNonKeyAttribute(root.getInfo().getRightJoinExpression(), right)) {
+          throw new KsqlException("Invalid join condition:"
+              + " stream-table joins require to join on the table's primary key.");
+        }
+      }
+
+      // stream-stream join detected: nothing to verify
+
+    } else {
+      if (right.getNodeOutputType() == DataSourceType.KSTREAM) {
+        throw new KsqlException("Invalid join order:"
+            + " table-stream joins are not supported; only stream-table joins.");
+      }
+
+      // table-table join detected
+
+      if (joinOnNonKeyAttribute(root.getInfo().getRightJoinExpression(), right)) {
+        throw new KsqlException("Invalid join condition:"
+            + " table-table joins require to join on the primary key of the right input table."
+        );
+      }
+
+      if (joinOnNonKeyAttribute(root.getInfo().getLeftJoinExpression(), left)) {
+        // after we lift the n-way join restriction, we should be able to support FK-joins
+        // at any level in the join tree, even after we add right-deep/bushy join tree support,
+        // because a FK-join output table has the same PK as its left input table
+        throw new KsqlException("Invalid join condition:"
+            + " foreign-key table-table joins are not supported.");
+      }
+    }
+  }
+
+  private boolean joinOnNonKeyAttribute(final Expression joinExpression,
+                                        final PlanNode node) {
+    if (!(joinExpression instanceof ColumnReferenceExp)) {
+      return true;
+    }
+    final ColumnReferenceExp simpleJoinExpression = (ColumnReferenceExp) joinExpression;
+
+    final String joinAttributeName = simpleJoinExpression.getColumnName().text();
+    final List<DataSourceNode> dataSourceNodes = node.getSourceNodes().collect(Collectors.toList());
+    final String singleAttributeKeyName;
+
+    // n-way join sub-tree (ie, not a leaf)
+    if (isInnerNode(node)) {
+      final DataSourceNode qualifiedNode;
+
+      if (simpleJoinExpression.maybeQualifier().isPresent()) {
+        final String qualifier = simpleJoinExpression.maybeQualifier().get().text();
+        final List<DataSourceNode> allNodes = dataSourceNodes.stream()
+            .filter(n -> n.getDataSource().getName().text().equals(qualifier))
+            .collect(Collectors.toList());
+
+        if (allNodes.size() != 1) {
+          // double check this
+          throw new KsqlException(String.format(
+              "Join qualifier '%s' could not be resolved (either not found or not unique).",
+              qualifier
+          ));
+        }
+        qualifiedNode = allNodes.get(0);
+      } else {
+        final List<DataSourceNode> allNodes = dataSourceNodes.stream()
+            .filter(n -> n.getSchema().columns().stream()
```

Comment:
Why not use `n.getSchema().findColumn()` to simplify this?

Reply:
From my understanding, as we have multiple source node, we first need to identify the correct source node base on the qualifier from the `joinExpression`, and can match the column name afterward . The `ColumnName` itself is just a plain `String` without qualifier and thus if we use `n.getSchema().findColumn()` we might find the column multiple times and/or in the wrong data source schema?

Reply:
I don't understand. Why is 
```
.filter(n -> n.getSchema().columns().stream()
    .map(column -> column.name().text()).collect(Collectors.toList())
    .contains(joinAttributeName))
```
(as is currently written in the PR) different from
```
.filter(n -> n.getSchema().findColumn(ColumnName.of(joinAttributeName)).isPresent())
```
?

Reply:
My bad... I was looking into the wrong part of the code...